### PR TITLE
More name hints omitting heuristics

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Omit parameter name hints for cases like the first two parameters from the screenshot:

Screenshot 2020-05-03 at 15 06 34

### CODE REVIEW

File: crates/ra_ide/src/inlay_hints.rs [hunk 1]
```diff
@@ -254,20 +255,57 @@ fn should_show_param_hint(
...
+    argument: &ast::Expr,
+    param_name: &str,
+) -> bool {
+    match sema.type_of_expr(argument).and_then(|t| t.as_adt()) {
+        Some(Adt::Enum(e)) => &camel_case_to_snake_case(e.name(sema.db).to_string()) == param_name,
+        _ => false,
+    }
+}
+
+fn camel_case_to_snake_case(s: String) -> String {
```
Comment (SomeoneToIgnore) [inline crates/ra_ide/src/inlay_hints.rs:287]: This is a copy-paste from `xtask`, `gen_syntax.rs`, `to_lower_snake_case` method and looks rather alien here, should I move it elsewhere?

Comment (edwin0cheng) [reply crates/ra_ide/src/inlay_hints.rs:287]: stdx ?

Comment (kiljacken) [reply crates/ra_ide/src/inlay_hints.rs:287]: We already indirectly depend on [heck](https://crates.io/crates/heck) through salsa, maybe we should just use that?

Comment (SomeoneToIgnore) [reply crates/ra_ide/src/inlay_hints.rs:287]: I like stdx for now, but we can consider heck again if something else pops up.

File: crates/ra_ide/src/inlay_hints.rs [hunk 2]
```diff
@@ -254,20 +256,43 @@ fn should_show_param_hint(
...
+    if is_enum_name_similar_to_param_name(sema, argument, param_name) {
+        return true;
+    } else {
+        let argument_string = if let Some(repr) = get_string_representation(argument) {
+            repr
+        } else {
+            return false;
+        };
+        let argument_string = argument_string.trim_start_matches('_');
+        argument_string.starts_with(param_name) || argument_string.ends_with(param_name)
```
Comment (edwin0cheng) [inline crates/ra_ide/src/inlay_hints.rs:273]:
```suggestion
    if is_enum_name_similar_to_param_name(sema, argument, param_name) {
        return true;
    }
    match get_string_representation(argument) {
        None => false,
        Some(repr) => {
            let argument_string = repr.trim_start_matches('_');
            argument_string.starts_with(param_name) || argument_string.ends_with(param_name)
        }
    }
```

**Comment by matklad** [comment]
Looking at the screenshot, I wonder if we also need a rule like "**always** show parameter hit in parmeter-per-line case", like this one:

image

File: crates/ra_ide/src/inlay_hints.rs [hunk 3]
```diff
@@ -254,20 +256,42 @@ fn should_show_param_hint(
...
+    }
+}
+
+fn is_enum_name_similar_to_param_name(
+    sema: &Semantics<RootDatabase>,
+    argument: &ast::Expr,
+    param_name: &str,
+) -> bool {
+    match sema.type_of_expr(argument).and_then(|t| t.as_adt()) {
+        Some(Adt::Enum(e)) => to_lower_snake_case(&e.name(sema.db).to_string()) == param_name,
```
Comment (matklad) [inline crates/ra_ide/src/inlay_hints.rs:282]: Another allocation avoiding options is something like

```rust
fn normalize(s: &str) -> impl Iterator<Item=char> {
    s.chars().map(|it| it.to_ascii_lowercase()).filter(|it| it != '_')
}
```

though, we still allocate at least once anyway, for the `Name`

**Comment by SomeoneToIgnore** [comment]
> always show parameter hit in parmeter-per-line case

That might be just another heuristics, not a separate option IMO.
I'm a bit afraid to add a lot of similar options since it might be hard to implement and maintain them all.
I like intellij-rust approach with "show only smart hints" checkbox doing all the magic.

**Comment by matklad** [comment]
Yeah, I meant this as an on by default rule, I don't think we need a checkbox for this

**Pull request merged**